Fix time-based statistics (peaks, incorrect values)

unfug-at-github opened the pull request:
Fixing issues with time-based statistics (peaks, incorrect values) and provide values for statistical function even if only a single input value exists

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The statistics functions has issues with the time based functions:

- Currently, time-based functions (average step / linear) don't compute a proper average over the specified time interval. Instead the average is only computed between the first and last change within the interval. This produces unexpected results when values are not changing over a longer period.
- Currently, the computation of the time-based functions is only triggered when input values change or values are dropping out of the age range. However, the actual average changes over time, because the time interval is shifting.

The changes will fix the above mentioned issues. To do so, they add a new parameter "refresh interval" that allows to updated the computed averages even if the input states are not changing.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #119738 #98262 #67627
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

unfug-at-github committed:
fixing issues with time-based statistics (peaks, incorrect values when values aren't changing) and provide values for statistical function even if only a single input value exists

unfug-at-github committed:
Update sensor.py, fixed spelling issues

unfug-at-github committed:
Update sensor.py, another typo

ThomDietrich commented:
Hello @unfug-at-github,
its generally not a good idea to mix so many changes in one PR.
Still, please give me some time to review, this looks like some good additions! Thanks for the contribution :)

unfug-at-github commented:
> Hello @unfug-at-github, its generally not a good idea to mix so many changes in one PR. Still, please give me some time to review, this looks like some good additions! Thanks for the contribution :)

I agree. I should have at least put the fixes for the average and the changes for the "compute with only one value" functions into different requests.

With the average functions it's a little harder to split things up, because these changes are somewhat depending on each other. 

Let me know if things need to be put into separate requests.

Zamtakk commented:
I could really use this fix from this PR, @ThomDietrich could you give an estimate on when you have time to approve this PR?

joostlek commented:
I think this PR should try to split up more, for me as a reviewer its not clear why we exactly change what since this PR tries to solve 4 bugs, so I can't discover if this is the proper way to solve the bug.

Also, there is a merge conflict.

unfug-at-github commented:
I have created a separate pull request that only addresses the issue of the spikes (#127268). I'll add further ones to reduce the scope of this one.

unfug-at-github commented:
Added another PR for the single value case #127305

unfug-at-github committed:
introducing a state buffer that is properly
synchronized, so that buffer updates no longer
interfere with statistical computations

Commit:
fixed issue with the refresh interval for expiring values
added test for sensor updates before database load finishes

unfug-at-github commented:
I created a new PR to fix the async issues in general #128796
